docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+  // error is returned if there are two active versions of the table
+  // descriptor.
+  rpc Publish(PublishRequest) returns (PublishResponse) {}
+}
+```
+
+When a node starts up it will register a `Watch` with the lease master
+in order to receive a stream of version updates for all
+descriptors. Each node will maintain a map from `<descriptorID,
+version>` to `<TableDescriptor, expiration, localRefCount>`. When the
+node needs to use a descriptor for an operation it will look in its
+local map. If the descriptor is not there it will be read from the KV
+store and a lease will be acquired by sending an `Acquire` message to
+the lease master. The node will maintain the invariant that a new
+reference to a descriptor will only be added to the most recent
+version. When the reference count for an older version falls to 0 the
+node will send a `Release` message to the lease master.
+
+The `Watch` responses will notify the node when a new version of a
+descriptor is available which it will then read from the KV store if
+the node has a lease for the descriptor.
+
+When performing a schema change operation, the modification to the
+table descriptor is performed using the `Publish` operation which will
+fail if the version is invalid (i.e. too old) or if there are already
+two active versions of the descriptor in the cluster. On success, the
+`Publish` operation will write the new descriptor to the KV store and
+notify all watches of the new descriptor which will cause those nodes
+to stop using the old descriptor for new operations. As existing
+operations complete they will release their local reference to the old
+descriptor which will cause the reference count to drop to 0 and
+result in the lease being released. When all of the leases for the old
+descriptor have been released, the lease master will send out a
+notification to the watches that the version is `DEAD`.
+
+Leases will be granted for a duration measured in minutes (we'll
+assume 5m for the rest of this doc, though experimentation may tune
+this number). A node will renew recently used descriptors before their
+leases expires and ensure that when handing out a descriptor for local
+usage the lease will not expire for at least `<leaseDuration>/2`
+minutes. When a node starts it will issue a `Release` message with the
+`all` field set to true to indicate it wants to release any leases
+held by its previous incarnation. This will ensure that a transient
+failure does not require the lease expiration to occur before allowing
+a schema modification. When a node dies permanently or becomes
+unresponsive (e.g. detached from the network) the lease master will
+have to wait for any leases that node held to expire. This will result
+in an inability to perform more than one schema modification step to
+the descriptors referred to by those leases. The maximum amount of
+time this condition can exist is the lease duration.
+
+If the lease master dies a new master will be "elected" within
```

Comment:
A new election will *start* within 20s; this process is not instantaneous. And while it is guaranteed that the first `CPut` to execute will complete the election, getting to that point may depend on a raft election which could require multiple rounds (especially in a small cluster where there is a higher probability that the lease master will also be the raft leader of a relevant range)